### ISSUE TITLE
SIGSEGV PhiNode::Ideal(PhaseGVN*, bool)+0x98

### ISSUE THREAD

**Issue by lordbuddha** [corretto/corretto-11#325]
### Describe the bug
JVM process dying with SIGSEGV.    Seems to be when doing internal compilation.

### To Reproduce
Its random...   A couple of times a day on very busy EC2 instances for one particular application out of 10.

### Expected behavior
I do not expect the JVM to crash with SIGSEGV.

### Screenshots
Nope.

### Platform information
    OS: Amazon Linux 2
    Version Corretto-11.0.19.7.1

### Additional context
We have only just managed to upgrade to Corretto 11 in production.   We were on Corretto 8 prior, with no problems.

Please find attached the er log and the compiler reply log.
[replay_pid10362.log](https://github.com/corretto/corretto-11/files/11472963/replay_pid10362.log)

[hs_err_pid10362.log](https://github.com/corretto/corretto-11/files/11472962/hs_err_pid10362.log)

**Comment by navyxliu** [comment]
hi, @lordbuddha , 
if it's random, do they always crash for the same reason? 

from my reading, your program crashed in a c2 compiler thread. it was compiling a method `nz.co.airnz.vibe.carina.operations.AddBookingSourceOperation::execute` and encounter a malform PhiNode. 

if you always see the same stacktrace, we will investigate what causes the compilation error. 
Thank you for providing the replay file.  To use the replay file, we also need to collect your class files. do you mind sharing classes files with us?

```
Current CompileTask:
C2:162085912 73500       4       nz.co.airnz.vibe.carina.operations.AddBookingSourceOperation::execute (255 bytes)

Stack: [0x0000ffff5413c000,0x0000ffff5433c000],  sp=0x0000ffff54337430,  free space=2029k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x45cbc8]  PhiNode::Ideal(PhaseGVN*, bool)+0x98
V  [libjvm.so+0xaf6b90]  PhaseIterGVN::transform_old(Node*)+0xc4
V  [libjvm.so+0xaf31cc]  PhaseIterGVN::optimize()+0x12c
V  [libjvm.so+0x5267a8]  Compile::Optimize()+0x148
V  [libjvm.so+0x527d08]  Compile::Compile(ciEnv*, C2Compiler*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0xa28
V  [libjvm.so+0x4409e4]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, DirectiveSet*)+0xbc
V  [libjvm.so+0x531990]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x2e4
V  [libjvm.so+0x532de8]  CompileBroker::compiler_thread_loop()+0x528
V  [libjvm.so+0xcdd3a0]  JavaThread::thread_main_inner()+0xe0
V  [libjvm.so+0xcdaef4]  Thread::call_run()+0x94
V  [libjvm.so+0xa9c798]  thread_native_entry(Thread*)+0xf8
C  [libpthread.so.0+0x722c]  start_thread+0xac
```

**Comment by lordbuddha** [comment]
The stacktrace is always the same.   The actual method being compiled is not, but have only seen it in a couple of places.    Will have a search for more now and will post after this.

The app is pretty big (20,000 classes).   Sort of a mess due to evolution over a long time..   I cannot share the classes with you without sign-off from up high...   Certainly not on a public forum.... 

That class/method AddBookingSourceOperation::execute is quite simple.   There is no deeply nested logic, just a few simple if statements...  The class is heavily injected though.... Using Google Guice am wondering if the injection using an older Guice could be an issue, we have an upgrade going live in a couple of days.
  
 (am on NZ time)

**Comment by lordbuddha** [comment]
A couple more

```
hs_err_pid26799.log:Current CompileTask:
hs_err_pid26799.log-C2:92883595 72943   !   4       nz.co.airnz.vibe.business.makebooking.MakeBookingSupport::addBookingDetails (191 bytes)
hs_err_pid26799.log-
hs_err_pid26799.log-Stack: [0x0000ffff5739e000,0x0000ffff5759e000],  sp=0x0000ffff57599430,  free space=2029k
hs_err_pid26799.log-Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
hs_err_pid26799.log-V  [libjvm.so+0x45cbc8]  PhiNode::Ideal(PhaseGVN*, bool)+0x98
hs_err_pid26799.log-V  [libjvm.so+0xaf6b90]  PhaseIterGVN::transform_old(Node*)+0xc4
hs_err_pid26799.log-V  [libjvm.so+0xaf31cc]  PhaseIterGVN::optimize()+0x12c
hs_err_pid26799.log-V  [libjvm.so+0x5267a8]  Compile::Optimize()+0x148
hs_err_pid26799.log-V  [libjvm.so+0x527d08]  Compile::Compile(ciEnv*, C2Compiler*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0xa28
hs_err_pid26799.log-V  [libjvm.so+0x4409e4]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, DirectiveSet*)+0xbc
hs_err_pid26799.log-V  [libjvm.so+0x531990]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x2e4
hs_err_pid26799.log-V  [libjvm.so+0x532de8]  CompileBroker::compiler_thread_loop()+0x528
hs_err_pid26799.log-V  [libjvm.so+0xcdd3a0]  JavaThread::thread_main_inner()+0xe0
hs_err_pid26799.log-V  [libjvm.so+0xcdaef4]  Thread::call_run()+0x94
hs_err_pid26799.log-V  [libjvm.so+0xa9c798]  thread_native_entry(Thread*)+0xf8
hs_err_pid26799.log-C  [libpthread.so.0+0x722c]  start_thread+0xac
hs_err_pid26799.log-
hs_err_pid26799.log-
hs_err_pid26799.log-siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x00010004a13c49b0
hs_err_pid26799.log-
--
hs_err_pid4381.log:Current CompileTask:
hs_err_pid4381.log-C2:160153410 73387       4       nz.co.airnz.vibe.carina.operations.AddBookingSourceOperation::execute (255 bytes)
hs_err_pid4381.log-
hs_err_pid4381.log-Stack: [0x0000ffff74197000,0x0000ffff74397000],  sp=0x0000ffff74392430,  free space=2029k
hs_err_pid4381.log-Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
hs_err_pid4381.log-V  [libjvm.so+0x45cbc8]  PhiNode::Ideal(PhaseGVN*, bool)+0x98
hs_err_pid4381.log-V  [libjvm.so+0xaf6b90]  PhaseIterGVN::transform_old(Node*)+0xc4
hs_err_pid4381.log-V  [libjvm.so+0xaf31cc]  PhaseIterGVN::optimize()+0x12c
hs_err_pid4381.log-V  [libjvm.so+0x5267a8]  Compile::Optimize()+0x148
hs_err_pid4381.log-V  [libjvm.so+0x527d08]  Compile::Compile(ciEnv*, C2Compiler*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0xa28
hs_err_pid4381.log-V  [libjvm.so+0x4409e4]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, DirectiveSet*)+0xbc
hs_err_pid4381.log-V  [libjvm.so+0x531990]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x2e4
hs_err_pid4381.log-V  [libjvm.so+0x532de8]  CompileBroker::compiler_thread_loop()+0x528
hs_err_pid4381.log-V  [libjvm.so+0xcdd3a0]  JavaThread::thread_main_inner()+0xe0
hs_err_pid4381.log-V  [libjvm.so+0xcdaef4]  Thread::call_run()+0x94
hs_err_pid4381.log-V  [libjvm.so+0xa9c798]  thread_native_entry(Thread*)+0xf8
hs_err_pid4381.log-C  [libpthread.so.0+0x722c]  start_thread+0xac
hs_err_pid4381.log-
hs_err_pid4381.log-
hs_err_pid4381.log-siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x00010003af5d13c0
hs_err_pid4381.log-
```

**Comment by navyxliu** [comment]
no worry.  
Since you have class files and the replay file, could you try to replay it on your side?
you can grab fastdebug build from here. 

https://downloads.corretto.aws/#/downloads?build=nightly&version=11&architecture=aarch64&platform=generic_linux&debugLevel=fastdebug

make sure you set classpath for it. 
```
java -XX:+UnlockDiagnosticVMOptions -XX:+ReplayCompiles -XX:+ReplayIgnoreInitErrors -XX:ReplayDataFile=replay_pid10362.log
```

this will try to recompile your method on your side. if it does trigger the error, it would have clearer stacktrace.

**Comment by lordbuddha** [comment]
Sorry no joy....    After it core'd a couple of times I ran it like this.....   (on a later replay file).

amazon-corretto-11.0.20.2.1-linux-aarch64-fastdebug/bin/java -XX:+UnlockDiagnosticVMOptions -XX:+ReplayCompiles -XX:+ReplayIgnoreInitErrors -XX:ReplayDataFile=replay_pid18488.log -cp "jetty-127_0_0_1-8080-vibe_war-_isisservices-any-3032993121617598894/webapp/WEB-INF/classes:jetty-127_0_0_1-8080-vibe_war-_isisservices-any-3032993121617598894/webapp/WEB-INF/lib/*:/opt/airnz/jvm/PR-A_VBE_07f46688cc5fd34c4/apps/jetty-distribution-9.4.49.v20220914/lib/*:/opt/airnz/jvm/PR-A_VBE_07f46688cc5fd34c4/lib/*"

But it failed

java.lang.NoClassDefFoundError: vbits/nz/co/airnz/vibe/ConfigImpl
java.lang.NoClassDefFoundError: vbits/nz/co/airnz/vibe/ConfigImpl
Error while parsing line 23753: Can't find holder klass

java.lang.NoClassDefFoundError: vbits/nz/co/airnz/vibe/ConfigImpl
Error while parsing line 23754: Can't find holder klass

java.lang.NoClassDefFoundError: org/apache/tomcat/util/digester/ArrayStack
java.lang.NoClassDefFoundError: vbits/nz/co/airnz/vibe/ConfigImpl

there is no way it will ever find vbits/nz/co/airnz/vibe/ConfigImpl because our code generates the source for that, compiles it with the eclipse compiler and loads it via a separate classloader on startup.    Yep, odd but has been working for us >10 years.

The reality is we have 20,000+ classes of our own and 300 jar files on the classpath....  

Not sure where we can go from here...    

I have a couple of changes that will go live Thursday NZ time....   One to give more stack and the other to remove the option "-XX:+UseStringDeduplication", no idea if either of these would have any affect on this issue, but they are differences in the way this app runs as compared to the others not having this issue.

**Comment by navyxliu** [comment]
@lordbuddha 
can you replace release build with fastdebug build and run? 
The crash report from debug build contains more accurate and detailed information. 
also pay a attention to preserve the coredump file on your system.

**Comment by lordbuddha** [comment]
@navyxliu  That is the only option I guess...  It will be about a month before we can do that.   Mainly because I am away for 3 weeks.

We don't currently generate coredumps...   I can allow that, but will not be able to supply...   Too much private data.

**Comment by lordbuddha** [comment]
OK...   On this now that I am back from leave.    We have had no similar crashes since I increased the stack size.  We were running with -Xss384k and had done so for years on Java 1.8.   Before I went on leave I bumped this to -Xss512k and the issue has gone.  So it seems on Java 11 the stack usage is greater for the same app and we were flying too close to the sun.

We use a reduced stack size because we run lots of threads and when on premise were a little short on RAM.   Now on AWS we do not really have to be so conservative.   Yep, we could reduce the threads, but that would be lots of time and effort.
